### ISSUE TITLE
Patches of noise on the edge of the generated atlas images

### ISSUE THREAD

**Issue by kvttt** [neel-dey/Atlas-GAN#3]
Hi @neel-dey,

First of all, thanks for making the code available. I tried out the model on my data set and everything turned out to be great. Except that during training I noticed that starting from around 10,000 iteration, strange-looking patches of noise started to appear.

![1](https://user-images.githubusercontent.com/99367900/202797364-63ef4bab-6473-4219-85ef-a2035fc24588.png)

I saved the volume as .nii file and observed that there seemed to be more than one patch.

![2](https://user-images.githubusercontent.com/99367900/202797515-fa2c6f40-e3ab-4853-a6ca-aac0f5f1a24a.png)

I was trying to figure out the problem on my own. Here's what I have done. I confirmed that my rough template to start with did not have strange patches. All my data were normalized to 0 to 1. 
I can't figure out what is wrong. Intuitively, the presence of the regularizer should be able to prevent voxels from walking that far to the corners. Plus, similarity loss should be able to penalize the generation of random patches of noise at places where they shouldn't be. I have noticed that in the script that `clip_bckgnd = False`. I wonder if setting this to `True` could help resolve this problem.

I would really appreciate if you could provide me with some insight on how this happened. Hope to hear back from you soon!

**Comment by neel-dey** [comment]
Hi @kvttt, thanks for reaching out! Yup, setting `clip_bckgnd=True` (which just zeroes out the voxels outside of some intensity threshold of the linear average input in training) should resolve this issue.

For context, those atlas voxels aren't being *moved* there by some deformation as they're being synthesized by the template generator. Surprisingly, almost all of the ablations and baselines tested in the paper can do weird things in the empty background (I've speculated that this is because the LNCC registration loss is scale-invariant and something like `LNCC + 0.1*L1` might do better), although it is definitely worse when GAN training is unstable. 

Just in case setting `clip_bckgnd=True` brings some artifacts into the *foreground*, try increasing `losswt_gp` slightly and/or decreasing `d_ch`. 

Good luck and let me know how it goes :)

**Comment by kvttt** [comment]
Hi @neel-dey, thanks for the clear explanation. I will adjust the hyperparameters you suggested and get back to you as soon as I have a result! Thanks!

**Comment by kvttt** [comment]
Hi @neel-day, I tried setting `clip_bckgnd=True`. I am seeing less noise visible from a coronal view but still some decent amount of noise from a sagittal view.  

<img width="652" alt="Screenshot 2022-11-22 at 11 55 14 AM" src="https://user-images.githubusercontent.com/99367900/203375545-8f8bed81-0720-4cf9-ad6b-0c8fbc70d890.png">

At this point do you think it would be a good idea to try replacing the LNCC loss with MSE or MAE loss and start from there?

**Comment by neel-dey** [comment]
Interesting. If you look at [these lines](https://github.com/neel-dey/Atlas-GAN/blob/4f30012d742432c9f565c1676a2fce1342a61e1f/src/networks.py#L377--L380), you'll see that it's basically just clipping `atlas_inputs` (which is the linear average input) with a predefined heuristic threshold of `1e-2`. I would just increase the threshold to something higher that basically only keeps the brain in the mask.

But trying LNCC + MAE should be really interesting! I wouldn't do MSE/MAE only because that typically doesn't do brain registration as well. Let me know how either option goes!

**Comment by kvttt** [comment]
Thanks for the reply! I will first go ahead and check what the clipping mask looks like on my linear average input and let you know if that is the root of the problem. I will also try LNCC + MAE tonight.

**Comment by kvttt** [comment]
I tried setting the threshold to `1e-1` and it is now working perfectly. Thank you for the insights and thanks for making such an amazing model available!

**Comment by neel-dey** [comment]
Fantastic! I'm very happy to hear that. Feel free to reach out again if you have any other questions.
